### PR TITLE
Changes wall leaning into a component, makes windows leanable

### PR DESCRIPTION
**Pull request by SmArtKar**
## About The Pull Request

Requires #85491 to be merged first, as otherwise leaning on directional windows looks incredibly goofy
You can now lean on both fulltile and directional windows, and its possible to easily extend this functionality to other objects. Dunno if this even deserves to be called a refactor so not marking it in any way.

## Why It's Good For The Game

Just makes sense that you'd be able to do it

## Changelog
:cl:
add: You can now lean on windows the same way you can lean on walls
fix: You no longer stop leaning on walls after clicking on anything
/:cl:

### CODE REVIEW

File: code/game/turfs/closed/walls.dm [hunk 1]
```diff
@@ -34,66 +32,9 @@
...
-		COMSIG_MOVABLE_TELEPORTING,
-		COMSIG_ATOM_DIR_CHANGE,
-	))
-	animate(src, 0.2 SECONDS, pixel_x = base_pixel_x, pixel_y = base_pixel_y)
-	remove_traits(list(TRAIT_UNDENSE, TRAIT_EXPANDED_FOV), LEANING_TRAIT)
-	update_fov()
-
 /turf/closed/wall/Initialize(mapload)
 	. = ..()
+	AddComponent(/datum/component/leanable, 11)
```
Comment (MrMelbert) [inline code/game/turfs/closed/walls.dm:37]: Note putting a component on literally every wall will probably do unhappy things to init time and memory usage

Keeping in mind lavaland is almost entirely walls this is probably like new 100,000 components

Comment (MrMelbert) [reply code/game/turfs/closed/walls.dm:37]: Relevant: #71008 , having a single element on mineral turfs added _two_ seconds to init time

I actually thought about doing this pr two days ago but I didn't think of a way to get around this problem, best I thought about is putting the procs on `/turf`, and having the only ways the procs are called via an element/component or via `/wall/mousedrop`

File: code/datums/components/leanable.dm [hunk 2]
```diff
@@ -0,0 +1,84 @@
...
+	visible_message(
+		span_notice("[src] leans against [lean_target]."),
+		span_notice("You lean against [lean_target]."),
+	)
+	RegisterSignals(src, list(
+		COMSIG_MOB_CLIENT_PRE_MOVE,
+		COMSIG_LIVING_DISARM_HIT,
+		COMSIG_LIVING_GET_PULLED,
+		COMSIG_MOVABLE_TELEPORTING,
+		COMSIG_ATOM_DIR_CHANGE,
```
Comment (MrMelbert) [inline code/datums/components/leanable.dm:69]: Also for a free fix while you're here, `COMSIG_ATOM_DIR_CHANGE` is fired even if an atom's dir doesn't ultimately change, so you could change this to use `COMSIG_ATOM_POST_DIR_CHANGE` and check `new_dir != old_dir` to make leaning a lot less jank overall

**Comment by SomeoneYouProbablyKnow** [comment]
Leaning on mechs soon tm

File: code/game/objects/structures/window.dm [hunk 3]
```diff
@@ -78,6 +78,20 @@
 	if (flags_1 & ON_BORDER_1)
 		AddElement(/datum/element/connect_loc, loc_connections)
 
+/obj/structure/window/mouse_drop_receive(atom/dropping, mob/user, params)
+	. = ..()
+	AddComponent(/datum/component/leanable, 11, same_turf = (flags_1 & ON_BORDER_1), lean_check = CALLBACK(src, PROC_REF(lean_check)))
+	dropping.base_mouse_drop_handler(src, null, null, params)
```
Comment (MrMelbert) [inline code/game/objects/structures/window.dm:84]: Include a comment explaining what's going on here so people understand

File: code/game/turfs/closed/walls.dm [hunk 4]
```diff
@@ -108,6 +48,11 @@
 		fixed_underlay = string_assoc_list(fixed_underlay)
 		underlays += underlay_appearance
 
+/turf/closed/wall/mouse_drop_receive(atom/dropping, mob/user, params)
+	. = ..()
+	AddComponent(/datum/component/leanable, 11)
+	dropping.base_mouse_drop_handler(src, null, null, params)
```
Comment (MrMelbert) [inline code/game/turfs/closed/walls.dm:54]: Ditto

**Comment by DrAmazing343** [comment]
another CLASSIC SMARTKAR SMASH HIT. thank you king...

**Comment by SmArtKar** [comment]
@Ghommie we good now

File: code/game/objects/structures/windows/window.dm [hunk 5]
```diff
@@ -90,6 +92,24 @@
...
 		AddElement(/datum/element/connect_loc, loc_connections)
 
+/obj/structure/window/mouse_drop_receive(atom/dropping, mob/user, params)
+	. = ..()
+	if (added_leaning)
+		return
+	/// For performance reasons and to cut down on init times we are "lazy-loading" the leaning component when someone drags their sprite onto us, and then calling dragging code again to trigger the component
+	AddComponent(/datum/component/leanable, 11, same_turf = (flags_1 & ON_BORDER_1), lean_check = CALLBACK(src, PROC_REF(lean_check)))
+	added_leaning = TRUE
+	dropping.base_mouse_drop_handler(src, null, null, params)
```
Comment (SyncIt21) [inline code/game/objects/structures/windows/window.dm:102]: Why are you calling this proc manually? It's not meant to be called by anyone.

File: code/game/objects/structures/windows/window.dm [hunk 6]
```diff
@@ -90,6 +92,24 @@
 	if (flags_1 & ON_BORDER_1)
 		AddElement(/datum/element/connect_loc, loc_connections)
 
+/obj/structure/window/mouse_drop_receive(atom/dropping, mob/user, params)
+	. = ..()
+	if (added_leaning)
```
Comment (SyncIt21) [inline code/game/objects/structures/windows/window.dm:97]:
```suggestion
	if (GetComponent(/datum/component/leanable))
```

And do the same for the other one and remove the var

File: code/game/objects/structures/windows/window.dm [hunk 7]
```diff
@@ -90,6 +92,24 @@
 	if (flags_1 & ON_BORDER_1)
 		AddElement(/datum/element/connect_loc, loc_connections)
 
+/obj/structure/window/mouse_drop_receive(atom/dropping, mob/user, params)
+	. = ..()
```
Comment (SyncIt21) [inline code/game/objects/structures/windows/window.dm:96]: If the parent proc isn't doing anything then it isn't nessassary

**Review by SyncIt21 — CHANGES_REQUESTED** [review]

File: code/game/objects/structures/windows/window.dm (hunk 6 above)
Comment (SmArtKar) [reply code/game/objects/structures/windows/window.dm:97]: GetComponent is generally considered ugly and should be avoided when possible

**Comment by SmArtKar** [comment]
Realized that since now this can be hooked up to movables, it needs a check for destruction/moving the object you're leaning on. In fact, it probably should've had that since the start - breaking a wall someone is leaning on doesn't stop it either

File: code/game/objects/structures/windows/window.dm (hunk 6 above)
Comment (SyncIt21) [reply code/game/objects/structures/windows/window.dm:97]: According to who? It's preferred to storing components as refs and is best for boolean checking in this case.

Just do a search text in VS code on how many places this proc is used before making such statements

File: code/game/objects/structures/windows/window.dm (hunk 7 above)
Comment (Ghommie) [reply code/game/objects/structures/windows/window.dm:96]: Well, it handles buckling, even if windows aren't something you can normally buckle to.

**Pull request merged**